### ISSUE TITLE
Effect of test-time batch size

### ISSUE THREAD

**Issue by Sxing2** [serendipity1122/Pre-trained-Model-Guided-Fine-Tuning-for-Zero-Shot-Adversarial-Robustness#1]
Hi, thanks for your work! It's very interesting. 

I came across an unexpected issue when I run the code. After adversarially finetuning the model and keeping it fixed, the robustness accuracy is heavily influenced by the batch size at test time. Using a batch size of 512 fares way better on perturbed images than when the batch size is 128 or 64, while the clean accuracy is relatively stable. In general, a larger batch size at test time leads to much higher robustness accuracy. 

But I can't locate the problem, do you guys have any clue?

**Comment by Sxing2** [comment]
*unexpected

**Comment by Sxing2** [comment]
Hi, I think I've found the problem. In the function `def multiGPU_CLIP` that calculates cosine similarity scores between a batch of image features and the text features, the feature vectors are double L2-normalised:

`img_embed, scale_text_embed = model(images, text_tokens, prompt_token)` This line outputs the image features that have been L2 normalised, and text features that have been L2 normalised and then scaled by a scalar `CLIP.logit_scale.exp()`. The purpose of this scalar is to 'sharpen' the scores.

`img_embed_norm = img_embed / img_embed.norm(dim=-1, keepdim=True)` This line L2 normalises again the image features, which is fine because the features should not change after being normalised once again.

`scale_text_embed_norm = scale_text_embed / scale_text_embed.norm(dim=-1, keepdim=True)` This line causes the problem because it L2 normalises the text features that have been normalised and scaled, thereby discarding the scale.

After removing the double normalisation, the robustness performance is stable w.r.t. test-time batch sizes, although the reason is still unclear. I think the reason might be that the perturbations are generated based on `def multiGPU_CLIP`, and discarding the scale in the function can lead to weak and unstable perturbations. 

Aside from this issue, this is still a piece of interesting work. Thanks again!
